File: subscription-will-not-end/python/expire_overdue_subscriptions.py
"""Expire WooCommerce Subscriptions whose end date has already passed.

A subscription with a schedule "end_date" is supposed to stop billing and move to
Expired on its own, driven by an Action Scheduler hook. If that hook was deleted,
never queued, or missed its run while the site was down, the subscription just sits
on Active (or On hold) forever with an end date in the past. This walks subscriptions
with a set end_date, checks whether that date has passed, and moves any overdue one to
Expired through the REST API, the same way the scheduled hook would have. It also
confirms with Stripe that there is no unexpected still-active off-session mandate on
an already-overdue subscription, since a store deleting a stale mandate is safer than
leaving it live once the subscription should be gone. Read only by default. Run on a
schedule.
"""
import os
import logging
from datetime import datetime, timezone

import stripe
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("expire_overdue_subscriptions")

stripe.api_key = os.environ["STRIPE_SECRET_KEY"]
WOO_URL = os.environ["WOO_STORE_URL"].rstrip("/")
AUTH = HTTPBasicAuth(os.environ["WOO_CONSUMER_KEY"], os.environ["WOO_CONSUMER_SECRET"])
GRACE_HOURS = int(os.environ.get("GRACE_HOURS", "6"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

OPEN_STATUSES = {"active", "on-hold", "pending-cancel"}
NO_END_DATE = "0000-00-00 00:00:00"


def parse_gmt(value):
    """Parse a WooCommerce *_date_gmt string into an aware UTC datetime, or None."""
    if not value or value == NO_END_DATE:
        return None
    text = value.replace("T", " ").replace("Z", "")
    return datetime.strptime(text, "%Y-%m-%d %H:%M:%S").replace(tzinfo=timezone.utc)


def intent_id_of(subscription):
    """The saved Stripe PaymentIntent id, from meta _stripe_intent_id or transaction_id."""
    for meta in subscription.get("meta_data") or []:
        if meta.get("key") == "_stripe_intent_id" and meta.get("value"):
            return meta["value"]
    tid = subscription.get("transaction_id")
    return tid if tid and tid.startswith("pi_") else None


def decide(subscription, now):
    """Pure decision: should this subscription be expired right now?

    subscription is a plain dict shaped like the WooCommerce Subscriptions REST API
    response (status, end_date_gmt, meta_data, transaction_id). now is an aware
    datetime, passed in so the function has no hidden clock and stays pure.
    """
    if subscription["status"] not in OPEN_STATUSES:
        return ("skip", "subscription is not in an open state")
    end_date = parse_gmt(subscription.get("end_date_gmt"))
    if end_date is None:
        return ("skip", "subscription has no end date, it renews until cancelled")
    if now < end_date:
        return ("skip", "end date has not arrived yet")
    overdue_hours = (now - end_date).total_seconds() / 3600
    if overdue_hours < GRACE_HOURS:
        return ("wait", "end date passed but still inside the grace window")
    return ("expire", f"end date passed {overdue_hours:.1f}h ago and is still open")


def get_subscription(subscription_id):
    r = requests.get(
        f"{WOO_URL}/wp-json/wc/v3/subscriptions/{subscription_id}", auth=AUTH, timeout=30
    )
    if r.status_code == 404:
        return None
    r.raise_for_status()
    return r.json()


def open_subscriptions():
    page = 1
    while True:
        r = requests.get(
            f"{WOO_URL}/wp-json/wc/v3/subscriptions",
            params={"status": "active,on-hold,pending-cancel", "per_page": 50, "page": page},
            auth=AUTH, timeout=30,
        )
        r.raise_for_status()
        batch = r.json()
        if not batch:
            return
        for subscription in batch:
            yield subscription
        page += 1


def cancel_stale_mandate(subscription):
    """Best-effort: cancel a Stripe PaymentIntent still sitting in a capturable state
    on a subscription we are about to expire. Never raises, since this is a courtesy
    cleanup and must not block the expiry itself.
    """
    intent_id = intent_id_of(subscription)
    if not intent_id:
        return
    try:
        intent = stripe.PaymentIntent.retrieve(intent_id)
        if intent.status in ("requires_capture", "requires_confirmation", "requires_action"):
            stripe.PaymentIntent.cancel(intent_id)
            log.info("Cancelled stale PaymentIntent %s on subscription %s", intent_id, subscription["id"])
    except stripe.error.StripeError as exc:
        log.warning("Could not check/cancel PaymentIntent %s: %s", intent_id, exc)


def mark_expired(subscription_id, reason):
    requests.put(
        f"{WOO_URL}/wp-json/wc/v3/subscriptions/{subscription_id}",
        json={"status": "expired"},
        auth=AUTH, timeout=30,
    ).raise_for_status()
    requests.post(
        f"{WOO_URL}/wp-json/wc/v3/subscriptions/{subscription_id}/notes",
        json={"note": f"Expired by the scheduled check: {reason}. "
                      f"The end date had passed but the subscription was still open."},
        auth=AUTH, timeout=30,
    ).raise_for_status()


def run():
    now = datetime.now(timezone.utc)
    expired = 0
    for subscription in open_subscriptions():
        action, reason = decide(subscription, now)
        if action != "expire":
            continue
        log.info("Subscription %s: %s. %s", subscription["id"], reason,
                  "would expire" if DRY_RUN else "expiring")
        if not DRY_RUN:
            cancel_stale_mandate(subscription)
            mark_expired(subscription["id"], reason)
        expired += 1
    log.info("Done. %d subscription(s) %s.", expired, "to expire" if DRY_RUN else "expired")


if __name__ == "__main__":
    run()

File: subscription-will-not-end/node/expire-overdue-subscriptions.js
/**
 * Expire WooCommerce Subscriptions whose end date has already passed.
 *
 * A subscription with a schedule "end_date" is supposed to stop billing and move to
 * Expired on its own, driven by an Action Scheduler hook. If that hook was deleted,
 * never queued, or missed its run while the site was down, the subscription just sits
 * on Active (or On hold) forever with an end date in the past. This walks
 * subscriptions with a set end_date, checks whether that date has passed, and moves
 * any overdue one to Expired through the REST API, the same way the scheduled hook
 * would have. It also does a best-effort check with Stripe to cancel a stale
 * PaymentIntent left on a subscription that should already be gone. Read only by
 * default. Run on a schedule.
 *
 * Guide: https://www.allanninal.dev/woocommerce/subscription-will-not-end/
 */
import Stripe from "stripe";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const GRACE_HOURS = Number(process.env.GRACE_HOURS || 6);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const OPEN_STATUSES = new Set(["active", "on-hold", "pending-cancel"]);
const NO_END_DATE = "0000-00-00 00:00:00";

export function parseGmt(value) {
  if (!value || value === NO_END_DATE) return null;
  const text = value.replace("T", " ").replace("Z", "") + "Z";
  const iso = text.replace(" ", "T");
  const date = new Date(iso);
  return Number.isNaN(date.getTime()) ? null : date;
}

export function intentIdOf(subscription) {
  for (const meta of subscription.meta_data || []) {
    if (meta.key === "_stripe_intent_id" && meta.value) return meta.value;
  }
  const tid = subscription.transaction_id;
  return tid && tid.startsWith("pi_") ? tid : null;
}

/**
 * Pure decision: should this subscription be expired right now?
 * `now` is a Date, passed in so the function has no hidden clock and stays pure.
 */
export function decide(subscription, now) {
  if (!OPEN_STATUSES.has(subscription.status)) {
    return ["skip", "subscription is not in an open state"];
  }
  const endDate = parseGmt(subscription.end_date_gmt);
  if (!endDate) {
    return ["skip", "subscription has no end date, it renews until cancelled"];
  }
  if (now < endDate) {
    return ["skip", "end date has not arrived yet"];
  }
  const overdueHours = (now - endDate) / 3600000;
  if (overdueHours < GRACE_HOURS) {
    return ["wait", "end date passed but still inside the grace window"];
  }
  return ["expire", `end date passed ${overdueHours.toFixed(1)}h ago and is still open`];
}

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

async function* openSubscriptions() {
  let page = 1;
  while (true) {
    const batch = await woo(`/subscriptions?status=active,on-hold,pending-cancel&per_page=50&page=${page}`);
    if (!batch || !batch.length) return;
    for (const subscription of batch) yield subscription;
    page++;
  }
}

async function cancelStaleMandate(subscription) {
  const intentId = intentIdOf(subscription);
  if (!intentId) return;
  try {
    const intent = await stripe.paymentIntents.retrieve(intentId);
    if (["requires_capture", "requires_confirmation", "requires_action"].includes(intent.status)) {
      await stripe.paymentIntents.cancel(intentId);
      console.log(`Cancelled stale PaymentIntent ${intentId} on subscription ${subscription.id}`);
    }
  } catch (err) {
    console.warn(`Could not check/cancel PaymentIntent ${intentId}: ${err.message}`);
  }
}

async function markExpired(subscriptionId, reason) {
  await woo(`/subscriptions/${subscriptionId}`, {
    method: "PUT",
    body: JSON.stringify({ status: "expired" }),
  });
  await woo(`/subscriptions/${subscriptionId}/notes`, {
    method: "POST",
    body: JSON.stringify({
      note: `Expired by the scheduled check: ${reason}. ` +
            `The end date had passed but the subscription was still open.`,
    }),
  });
}

export async function run() {
  const now = new Date();
  let expired = 0;
  for await (const subscription of openSubscriptions()) {
    const [action, reason] = decide(subscription, now);
    if (action !== "expire") continue;
    console.log(`Subscription ${subscription.id}: ${reason}. ${DRY_RUN ? "would expire" : "expiring"}`);
    if (!DRY_RUN) {
      await cancelStaleMandate(subscription);
      await markExpired(subscription.id, reason);
    }
    expired++;
  }
  console.log(`Done. ${expired} subscription(s) ${DRY_RUN ? "to expire" : "expired"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((e) => { console.error(e); process.exit(1); });
}
